File: examples/clj/test/com/rpl/agent/basic/aggregation_agent_test.clj
(ns com.rpl.agent.basic.aggregation-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.aggregation-agent :refer [AggregationAgentModule]]))

(deftest aggregation-agent-test
  (System/gc)
  (testing "AggregationAgent example produces expected fan-out/fan-in behavior"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc AggregationAgentModule {:tasks 2 :threads 2})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        AggregationAgentModule))
            agent   (aor/agent-client manager "AggregationAgent")]

        (testing "processes data in chunks and aggregates results"
          (let [result (aor/agent-invoke agent
                                         {:data       (range 1 13) ; [1 2 3 ...
                                          ; 12]
                                          :chunk-size 4})]
            ;; Verify final result structure
            (is (= 12 (:total-items result)))
            (is (= 3 (:chunks-processed result)))

            ;; Verify total sum calculation
            ;; Each item is squared, so sum = 1² + 2² + ... + 12² = 650
            (is (= 650 (:total-sum result)))

            ;; Verify chunk results structure
            (let [chunk-results (:chunk-results result)]
              (is (= 3 (count chunk-results)))
              (is (every? #(contains? % :original-chunk) chunk-results))
              (is (every? #(contains? % :processed-chunk) chunk-results))
              (is (every? #(contains? % :chunk-sum) chunk-results))

              ;; Verify chunk sums add up to total
              (is (= 650 (reduce + (map :chunk-sum chunk-results))))

              ;; Verify chunk processing (each original value squared)
              (is (= [1 4 9 16] (:processed-chunk (first chunk-results))))
              (is (= [25 36 49 64] (:processed-chunk (second chunk-results))))
              (is (= [81 100 121 144]
                     (:processed-chunk (nth chunk-results 2)))))))

        (testing "handles different chunk sizes correctly"
          (let [result (aor/agent-invoke agent
                                         {:data       (range 1 8) ; [1 2 3 4 5 6
                                          ; 7]
                                          :chunk-size 3})]
            (is (= 7 (:total-items result)))
            (is (= 3 (:chunks-processed result))) ; [1,2,3], [4,5,6], [7]
            ;; Sum = 1² + 2² + ... + 7² = 140
            (is (= 140 (:total-sum result)))))))))

File: examples/java/src/test/java/com/rpl/agent/basic/AggregationAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.Test;

/**
 * Test class for AggregationAgent demonstrating fan-out/fan-in aggregation patterns.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>aggStartNode distributing work to multiple parallel processors
 *   <li>aggNode collecting and combining results from multiple executions
 *   <li>Built-in LIST_AGG aggregator functionality
 *   <li>Fan-out/fan-in execution patterns with different chunk sizes
 * </ul>
 */
public class AggregationAgentTest {

  @Test
  public void testBasicAggregation() throws Exception {
    // Tests basic aggregation functionality with simple data
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      AggregationAgent.AggregationModule module = new AggregationAgent.AggregationModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("AggregationAgent");

      // Test with simple data set
      List<Integer> testData = new ArrayList<>();
      for (int i = 1; i <= 10; i++) {
        testData.add(i); // [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
      }

      Map<String, Object> request = new HashMap<>();
      request.put("data", testData);
      request.put("chunkSize", 3);

      Map<String, Object> result = (Map<String, Object>) agent.invoke(request);

      assertNotNull("Result should not be null", result);
      assertEquals("Should process all 10 items", 10, result.get("totalItems"));
      assertEquals("Should create 4 chunks", 4, result.get("chunksProcessed"));

      // Expected: chunks [1,2,3], [4,5,6], [7,8,9], [10]
      // Squared: [1,4,9], [16,25,36], [49,64,81], [100]
      // Sums: 14 + 77 + 194 + 100 = 385
      assertEquals("Total sum should be correct", 385, result.get("totalSum"));
    }
  }
}
